engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java

```diff
@@ -241,23 +243,29 @@ public void resetModuleSources(File path) {
    * @param edits the edits to apply.
    * @return an object for computing the changed IR nodes.
    */
-  public Optional<Changeset<Rope>> modifyModuleSources(File path, List<model.TextEdit> edits) {
+  public Optional<ChangesetBuilder<Rope>> modifyModuleSources(File path, List<model.TextEdit> edits) {
     Optional<Module> moduleMay = context.getModuleForFile(path);
-    if (!moduleMay.isPresent()) {
+    if (moduleMay.isEmpty()) {
+      logger.severe("Failed to get module for file: " + path);
       return Optional.empty();
     }
     Module module = moduleMay.get();
-    if (module.getLiteralSource() == null) {
+    try {
+      module.getSource();
+    } catch (IOException e) {
+      logger.log(Level.SEVERE, "Failed to get source for file: " + path, e);
       return Optional.empty();
     }
-    Changeset<Rope> changeset =
-        new Changeset<>(
+    ChangesetBuilder<Rope> changesetBuilder =
+        new ChangesetBuilder<>(
             module.getLiteralSource(),
             module.getIr(),
             TextEditor.ropeTextEditor(),
             IndexedSource.RopeIndexedSource());
     Optional<Rope> editedSource = JavaEditorAdapter.applyEdits(module.getLiteralSource(), edits);
-    editedSource.ifPresent(module::setLiteralSource);
-    return Optional.of(changeset);
+    editedSource.ifPresentOrElse(
+        module::setLiteralSource,
+        () -> logger.severe("Failed to apply edits for file: " + path + ", " + edits));
```

Comment:
There's a lot of severe-level logs in this file. How many of them can be caused by external issues, and how many are down to programming bugs?

Reply:
All of them signal about logical issues in the code (aka bugs). I added logging because it should help to spot the source of an error in the future

Reply:
Sure, but traditionally we critically fail on programming bugs, rather than trying to continue in a broken state.

Reply:
Agree, I forgot that we already have a set of custom exceptions that we use in the execution service. I'll replace the logging with them :+1: